Added ESLint and Prettier Format for Nest Frontend, Added Husky Setup

tsu-ki opened the pull request:
Added ESLint, Prettier format and completed Husky setup to run pre-commit tests. 
Added Instructions for all of them in README.md file.
https://github.com/OWASP/Nest/issues/78

<img width="722" alt="Screenshot 2024-11-26 at 9 18 46 PM" src="https://github.com/user-attachments/assets/f71d2772-1344-46c5-b408-55b0494fd9b6">
<img width="899" alt="Screenshot 2024-11-26 at 10 13 28 PM" src="https://github.com/user-attachments/assets/f3266651-f08a-49e2-b365-6a6c98d01d11">
<img width="899" alt="Screenshot 2024-11-26 at 10 13 53 PM" src="https://github.com/user-attachments/assets/3379c6c8-293a-4836-b693-f35c5edc6802">

arkid15r commented on README.md:
````diff
@@ -1,22 +1,152 @@
+### Configure Husky
+To set up Husky for pre-commit hooks, run:
+
+```bash
+npx husky install
+```
+Then create a pre-commit hook:
+
+```bash
+npx husky-init
+```
+
+### Pre-commit Hook Script
+Create or update the `.husky/pre-commit` file with the following content:
+
+```bash
+#!/bin/sh
+# .husky/pre-commit
+echo "Running pre-commit hook..."
+echo "Running lint-staged..."
+export NODE_OPTIONS="--max-old-space-size=4096"
+npx lint-staged
+
+# Optional: Run ESLint and Prettier directly for debugging
+eslint 'frontend/**/*.{ts,tsx}' --fix --config eslint.config.js >> .husky/log.txt 2>&1
+prettier --write 'frontend/**/*.{ts,tsx}' >> .husky/log.txt 2>&1
+```
+
+
+### Linting and Formatting Commands
+You can manually run ESLint and Prettier to check your setup:
+   - To lint your code:
+```bash
+npx eslint 'frontend/**/*.{js,jsx,ts,tsx}' --fix --config eslint.config.js
+```
+
+  - To format your code with Prettier:
+```bash
+npx prettier --write 'frontend/**/*.{js,jsx,ts,tsx}'
+```
````
The setup part needs to be automated. The linter setup documentation is longer than the rest of the project setup now. Our goal here is to keep the new developers onboarding process as easy as possible. Everything should be pre-configured and installed automatically -- I suggest wrapping it with Makefile commands and avoid copy/paste instructions

tsu-ki replied:
I believe that would make more sense and would create less confusion. I'll go study how makefile commands work and get this implemented aswell. thank you for the reviews! If there's anything else, please do let me know

arkid15r commented on README.md:
````diff
@@ -1,22 +1,152 @@
+     git clone https://github.com/owasp/nest
+     ```
+
+2. **Create Environment File**:
````
Could you keep `1.` everywhere for auto- numbering

tsu-ki replied:
sure thing, i'll change it right away

arkid15r commented on package.json:
This file and other config files belong in `frontend/` directory, right?

tsu-ki replied:
the linting and prettier format will work only for files present in `frontend/`. But the config files itself aren't located inside the frontend directory.

tsu-ki replied:
sure thing, I'll make the required changes

arkid15r reviewed (commented):
The workflow works for me. What we need to improve is integration into the current setup by minimizing the required setup/use manual steps. This impacts the developer's  experience a lot.

arkid15r commented on frontend/package.json:
```diff
@@ -4,10 +4,14 @@
   "type": "module",
   "scripts": {
-    "dev": "vite --host --port 3000",
+    "dev": "vite --port 3000",
```
```suggestion
    "dev": "vite --host --port 3000",
```

arkid15r commented on README.md:
````diff
@@ -1,22 +1,134 @@
+Make sure you have the necessary dependencies installed:
+
+```bash
+npm install --save-dev eslint prettier husky lint-staged eslint-plugin-react eslint-plugin-react-hooks @typescript-eslint/eslint-plugin @typescript-eslint/parser eslint-config-prettier eslint-plugin-prettier
````
This should be installed automatically upon docker stack run, right?
if `npm install` takes care of this we can remove it from here.

harsh3dev replied:
When you would update the `package.json`  then these dependencies will automatically get installed right?
So I feel and you would see the same that `npm install` must install all these dependencies rather than explicitly doing it.

arkid15r replied:
> umm, my thought process behind putting these commands in readme was to provide users with a short section for commands and required dependencies, as if someone had problems setting up the project, they could simply follow the mentioned instructions. Please do suggest what feels best here

This would suit better for a troubleshooting section (when we have one). For the initial setup the fewer instructions the better. Ideally one command per a setup step.

tsu-ki replied:
@harsh3dev @arkid15r yes, after `npm install` all the dependencies will be automatically installed. On running `npm run prepare` husky will be installed and `.husky/pre-commit` file will be created with the required content already present inside. 
We'll only have to run a single husky command to execute linting aswell as prettier formatting.
I mentioned the ESLint and Prettier commands in the `README.md` so that if someone wanted to test run either of them, they could easily do it without having to look for the right command.

arkid15r commented on README.md:
````diff
@@ -1,22 +1,134 @@
+### Pre-commit Hook Script
+Ensure the `.husky/pre-commit` file consists of the following content:
+
+```bash
````
Would it be a good idea to put it to the `.husky/pre-commit` file in the code base instead of creating it locally for everyone?

tsu-ki replied:
@arkid15r  umm, I had problem setting up husky with the commands you initially suggested in the issue, as they were deprecated. I'm not sure if putting the husky inside the code base would be the best choice as versions might change again. Can we ask for reviews from the other collaborator aswell?

harsh3dev replied:
Can we use `npx husky add .husky/pre-commit "npm run lint"` this one creates the husky pre-commit hook.

tsu-ki replied:
@harsh3dev running `npm run prepare` would automatically install husky and create `.husky/pre-commit` hook with the required contents

<img width="1440" alt="Screenshot 2024-12-03 at 7 35 02 AM" src="https://github.com/user-attachments/assets/1a20811b-1433-4fb8-b39e-5424237a625b">
<img width="1440" alt="Screenshot 2024-12-03 at 7 35 24 AM" src="https://github.com/user-attachments/assets/9fcbc6cb-af16-4c63-a1fd-e56747f7b466">

@arkid15r this is how Linting structure/commands will be displayed `README.md`

arkid15r commented on README.md:
```diff
@@ -1,22 +1,134 @@
+# .husky/pre-commit
+echo "Running pre-commit hook..."
+echo "Running lint-staged..."
+export NODE_OPTIONS="--max-old-space-size=4096"
```
Could you add more details on why this is needed?

tsu-ki replied:
@arkid15r the `echo` lines were basically for logging/printing the start of the pre-commit process, same for `lint-staged`.  `export NODE_OPTIONS="--max-old-space-size=4096"` this was for allocating a maximum of 4GB (4096MB) heap memory so that heavier tasks could run easily without running out of memory. I believe this isn't a necessity right now and could be removed if wanted, I added it as a safe measure

arkid15r commented on README.md:
````diff
@@ -1,22 +1,134 @@
+You can manually run ESLint and Prettier to check your setup:
+   - To lint your code:
+```bash
+npx eslint . --fix --config eslint.config.js
+```
+
+  - To format your code with Prettier:
+```bash
+npx prettier --write .
````
Ideally we want to have a single command, maybe another Makefile target if possible to use it as another step for `make pre-commit` command.

It should be easy for a developer to make the code to comply with the expected requirements/style by using just a single command.

tsu-ki replied:
@arkid15r I put these commands so that user could manually check if their lint and prettier format setup was correct. Originally, running the `husky pre-commit` command will perform both of these tasks

arkid15r replied:
Let's think on how to simplify it after merging this PR

arkid15r reviewed (commented):
Thanks for moving it to `frontend`

Now let's focus on simplifying things to everyone:

arkid15r commented:
@tsu-ki could you address the rest of comments and request re-review when ready?
Thank you!

tsu-ki commented:
> @tsu-ki could you address the rest of comments and request re-review when ready? Thank you!

I apologise for not addressing the concerns before, I've replied to each of the review and would be glad if you or Harsh could suggest any more probable changes. thanks!

arkid15r reviewed (approved):
@tsu-ki that's a great piece of code quality improvement work!
I've merged some conflicts and run lint/format commands on the existing files.
Let's see how it works and improve it incrementally.

Thanks for adding this!
